analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs

```diff
@@ -0,0 +1,609 @@
+﻿using System;
+
+namespace Tests.Diagnostics
+{
+    public class Program
+    {
+        private object field = null;
+        private static object staticField = null;
+
+        public void NotCompliantCases(object o, Exception e)
+        {
+            o.ToString(); // FIXME Non-compliant {{Refactor this method to add validation of parameter 'o' before using it.}}
+
+            Bar(o); // Compliant, we care about dereference only
+
+            throw e; // FIXME Non-compliant
+        }
+
+        public void Bar(object o) { }
+
+        protected void NotCompliantCases_Nonpublic(object o)
+        {
+            o.ToString(); // FIXME Non-compliant
+        }
+
+        private void CompliantCases_Private(object o)
+        {
+            o.ToString(); // Compliant, not public
+        }
+
+        protected internal void NotCompliantCases_ProtectedInternal(object o)
+        {
+            o.ToString(); // FIXME Non-compliant
+        }
+
+        internal void CompliantCases_Internal(object o)
+        {
+            o.ToString(); // Compliant, not public
+        }
+
+        public void CompliantCases(bool b, object o1, object o2, object o3, object o4, Exception e)
+        {
+            if (o1 != null)
+            {
+                o1.ToString(); // Compliant, we did the check
+            }
+
+            o2 = o2 ?? new object();
+            o2.ToString(); // Compliant, we coalesce
+
+            if (o3 == null)
+            {
+                throw new Exception();
+            }
+
+            o3.ToString(); // Compliant, we did the check
+
+            if (e != null)
+            {
+                throw e; // Compliant
+            }
+
+            o4?.ToString(); // Compliant, conditional operator
+
+            b.ToString(); // Compliant, bool cannot be null
+
+            object v = null;
+            v.ToString(); // Compliant, we don't care about local variables
+
+            field.ToString(); // Compliant
+
+            Program.staticField.ToString(); // Compliant
+        }
```

Comment:
Some other interesting scenarios, which don't seem to be currently covered:
- transitivity of nullable and non-nullable constraints on comparison => given that `o1 != null`, if you compare it against an object `o5` with no constraints, constraints propagate from `o1` to `o5`;
- transitivity of  nullable and non-nullable constraints on assignment => same as case above, but with an assignment: if you assign `o1` to `o5`, constraints propagage from `o1` to `o5`.

Reply:
Those are really good test cases. Added them.